### ISSUE TITLE
Youtube Center causes Firefox memory usage to grow to 3GB and then crash in 38.0.5

### ISSUE THREAD

**Issue by keLston** [YePpHa/YouTubeCenter#1875]
In 38.0.1 of Firefox, I was noticing that watching YT videos was causing Firefox memory usage to grow and peak at 2GB and the browser would lag and the only way to fix it was to restart Firefox. When 38.0.5 was released, this only got worse since the memory usage would grow even further to 3GB and then the browser would subsequently crash.

I narrowed it down to Youtube Center as the cause. It wasn't happening with Firefox running in safemode so I slowly enabled addons until it started reoccurring and lo and behold, it didn't reoccur until YTC was enabled.

**Comment by Piter432** [comment]
Confirmed. Screw up bugs, glitched functions of YTC or different YTC things caused by YouTube/Google changes in almost all YouTube pages. Memory usage is very important now. Or maybe Mozilla will soon release Firefox 39.0 with enabled e10s by default, but I don't remember is YTC compatible with Firefox'es e10s.

**Comment by ErdeFB** [comment]
@keLston Did you also test with only YTC running with no other addons?
Just checking in case it is a problem with two addons running together rather than just YTC doing it itself.

@Piter432 YTC should be compatible with e10s.

**Comment by keLston** [comment]
@ErdeFB - Alright. I just tried it. Disabled everything except Greasemonkey and Youtube Center. I did nothing except launch Firefox to check it's initial mem usage (290MB) and then on only one tab I went to Youtube. In 6 videos, I was able to balloon the memory usage to 1.5GB.

**Comment by Piter432** [comment]
@ErdeFB So "should" or "is" compatible? ;>

**Comment by ErdeFB** [comment]
I seem to recall seeing the author commenting on this issue tracker that he had made it compatible. When I tested it last week, it worked. But with high memory consumption.

**Comment by Yonezpt** [comment]
@Piter432 YTC is E10 compatible https://github.com/YePpHa/YouTubeCenter/issues/1824#issuecomment-103109280

**Comment by Piter432** [comment]
@Yonezpt Oh, I forgot that :P

**Comment by DannyPockets** [comment]
I too have been experiencing issues with memory. The memory usage would skyrocket to 1GB after a few videos. Also firefox will freeze most the time when I click the play/pause button, click on the player settings or drag the video bar. It will freeze for a few seconds then continue to work but happens way too often now. I've tried disabling custom buffers but it doesn't help one bit.

**Comment by leonard84** [comment]
I can confirm the memory leak in 38.0.5 with YTC Devel 525

**Comment by Morphello** [comment]
Adding to this. Waterfox 38.1 (64bit firefox build) and Firefox 39. Same profile, all addons enabled. 
Fresh start, YTC 525 disabled, 15 videos opened, watched, then tabs closed: 533mb memory.
Close, reopen, YTC 525 enabled, 15 videos opened, watched, then tabs closed: 1735mb memory.

**Comment by bottleblue** [comment]
latest version still has memory leaks

**Comment by rdamiano** [comment]
Same for me. FF38.0.5 on Windows 7. The UI and webpage in general becomes incredibly laggy as the video plays/buffers with both the Flash and HTML players.

This is easily the biggest issue, YTC is unusable for all post-FF38 users.

**Comment by coldblackice** [comment]
@rdamiano What other plugins have you moved to, if you wouldn't mind sharing?

**Comment by coldblackice** [comment]
I wonder if forcing a garbage-collection within Firefox would address this? Surely there must be a way to trigger the release of the memory.

Another possibility is a third-party program called Process Lasso, which is able to force programs to trim their usage back (particularly when they've been overzealous with commits).

**Comment by Morphello** [comment]
I should have mentioned in my write up. Forcing GC actually adds a few megabytes to the memory usage.

**Comment by rdamiano** [comment]
@coldblackice For the time being I'm using Youtube+. Still prefer Youtube Center, though, especially the way it handled thumbnail previews.

**Comment by keLston** [comment]
@coldblackice No, I had tried that. Forcing garbage collection lowers memory usage by only a few MB while this problem is causing a hold and growth of hundreds of MB. Further, forcing the gc also failed to hold back the inevitable crash or the very notable performance lag.

**Comment by rdamiano** [comment]
Not to be demanding, but I see random commits and activity by @YePpHa and other contributors, just wondering if this is being worked on? No matter what else is fixed/added, this issue is a dealbreaker and makes YTC unusable for up-to-date Firefox users.

**Comment by MamiyaOtaru** [comment]
after watching a couple vids and closing the tabs, in about:memory is

│  ├───31.67 MB (05.07%) -- top(none)/ghost
│  │   ├──16.62 MB (02.66%) -- window(https://www.youtube.com/watch?v=-b*******)
│  │   │  ├───6.61 MB (01.06%) ++ js-compartment(https://www.youtube.com/)
│  │   │  ├───6.46 MB (01.03%) ++ dom
│  │   │  └───3.56 MB (00.57%) ++ (2 tiny)
│  │   ├──15.04 MB (02.41%) -- window(https://www.youtube.com/watch?v=-L*******)
etc.

They persist after GC and emptying caches (RAMBack extension).  With the Youtube Center Greasemonkey script disabled, after closing the tab and waiting a bit / clearing cache there is no such top(none)/ghost entry.  With it enabled, there an entry for every youtube tab closed under top(none)/ghost.

**Comment by coldblackice** [comment]
@MamiyaOtaru Interesting. Out of curiosity, how did you narrow it down to that out? What method/tools are you using?

**Comment by rdamiano** [comment]
I'm assuming this isn't something that can be fixed client-side, right? Still waiting for a dev build with this issue fixed...

**Comment by rdamiano** [comment]
Is this issue still present in FF 39? I'm assuming it didn't just go away and so far all of the commits have been for other very minor bugs...

**Comment by keLston** [comment]
@rdamiano Unfortunately, I couldn't say. Since I reported this and we haven't had an official response or fix yet, I haven't gone back to this addon. I couldn't deal with having my browser crash every 20 minutes.

**Comment by rdamiano** [comment]
Yeah, same. I don't understand why this isn't getting more attention as it makes the add-on absolutely useless for Firefox users (and maybe Chrome now, too https://www.reddit.com/r/YouTubeCenter/comments/3b2c3d/anyone_else_getting_lagged_up_from_youtubecenter/ct6q2xd).

Or maybe it's not getting ignored, we can't tell. It'd be great if we could just get some type of acknowledgement from @YePpHa or another dev.

**Comment by rdamiano** [comment]
Can anyone report if this is still an issue with the latest YTC in at least FF 40.0?

**Comment by Morphello** [comment]
Enabled it with my previous settings just to check. Checked Flash and HTML5. Both blow out the ram quickly still.

**Comment by rdamiano** [comment]
Thanks for checking it out!

I feel like this issue isn't necessarily being ignored, it's just that it's so old and with the issue tracker being constantly flooded with other issues, @YePpHa doesn't even know that his plugin is unusable on an entire browser due to this bug.

**Comment by keLston** [comment]
There's 4 major browsers to support. IE, Firefox, Safari, and Chrome. It would be absurd to completely not realize your addon doesn't work for one of them for 3 months.

**Comment by rdamiano** [comment]
@keLston It works, it just lags to the point of near-unresponsiveness after some time. The lack of acknowledgement of this particular issue leads me to believe @YePpHa isn't aware of its severity.

**Comment by keLston** [comment]
If your addon inherently causes the product it is an addon to to crash, then that is a very loose definition of working unless the designed functionality of the addon is to make the product crash.

**Comment by MamiyaOtaru** [comment]
@Coldblackice I was just looking for things in the script that referred to the browser chrome itself and not the page contained therein.  As the chrome does not go away on a page refresh or unload, the script remains in memory

No tools other than watching the memory with that setTimeout commented out and with it as is.
